src/logics/ArithLogic.cc

```diff
@@ -401,15 +410,14 @@ lbool ArithLogic::arithmeticElimination(vec<PTRef> const & top_level_arith, Subs
         PTRef lhs = logic.getPterm(eq)[0];
         PTRef rhs = logic.getPterm(eq)[1];
         PTRef polyTerm = lhs == logic.getZeroForSort(logic.getSortRef(lhs)) ? rhs : logic.mkMinus(rhs, lhs);
-        assert(logic.isLinearTerm(polyTerm));
         if (logic.isLinearFactor(polyTerm)) {
-            auto [var, c] = logic.splitTermToVarAndConst(polyTerm);
+            auto [var, c] = logic.splitPolyTerm(polyTerm);
             auto coeff = logic.getNumConst(c);
             poly.addTerm(var, std::move(coeff));
         } else {
-            assert(logic.isPlus(polyTerm));
+            assert(logic.isPlus(polyTerm) || logic.isTimesNonlin(polyTerm));
             for (PTRef factor : logic.getPterm(polyTerm)) {
-                auto [var, c] = logic.splitTermToVarAndConst(factor);
+                auto [var, c] = logic.splitPolyTerm(factor);
```

Comment:
Nonlinear multiplication should be handled as a factor, not as a sum, no? I think this is not correct.